### ISSUE TITLE
Incorrect KL Loss Sign Causes Policy Training Collapse

### ISSUE THREAD

**Issue by OVentoFAZ** [qiancheng0/ToolRL#52]
### Description

<img width="976" height="700" alt="Image" src="https://github.com/user-attachments/assets/9a0496b5-ae83-4a35-80de-a7894c512243" />
<img width="658" height="388" alt="Image" src="https://github.com/user-attachments/assets/33da9383-0471-4a7c-9b55-d9094b4954ad" />
<img width="967" height="694" alt="Image" src="https://github.com/user-attachments/assets/543728af-ca50-4f98-a14e-fc6c95430e7c" />
<img width="961" height="367" alt="Image" src="https://github.com/user-attachments/assets/bde359b4-0398-4026-ab28-d531d3042d7f" />
<img width="943" height="679" alt="Image" src="https://github.com/user-attachments/assets/cdfa881e-0310-4762-b9af-03483165b1a9" />
<img width="961" height="373" alt="Image" src="https://github.com/user-attachments/assets/62e9f6e8-689f-42cf-9525-9fbe1ab85a6c" />
<img width="952" height="679" alt="Image" src="https://github.com/user-attachments/assets/d1958167-dde2-42d9-8b2d-b6ceae3078d5" />
<img width="961" height="376" alt="Image" src="https://github.com/user-attachments/assets/30a0f529-ffcd-4e31-a9d4-842df0d5b86a" />

While using this project for GRPO training experiments, I noticed that when `use_kl_loss=True`, the training becomes unstable and eventually collapses. After investigation, I found a potential issue with the KL loss sign in the policy loss calculation.

### Location

**File**: `verl/workers/actor/dp_actor.py`  
**Line**: 267

```python
# Current code
policy_loss = policy_loss - kl_loss * self.config.kl_loss_coef
```

#### 1. Theoretical Background

In PPO's objective function, we aim to:
- Maximize: `E[advantage * ratio] - entropy_coeff * entropy - kl_coeff * KL(π_new || π_ref)`

Converting to minimization loss:
```
policy_loss = pg_loss - entropy_loss * entropy_coeff + kl_loss * kl_loss_coef
```

**Note**: The KL term should be **positive** (`+`), because we want to minimize KL (adding KL in the loss is equivalent to subtracting it in the maximization objective).

#### 2. Potential Issue

The current code uses a negative sign (`-`):
```python
policy_loss = policy_loss - kl_loss * self.config.kl_loss_coef
```

This might cause:
- When KL is large (policy diverges from reference), the loss decreases instead
- The optimizer may update the policy in the direction of increasing KL
- The policy could rapidly diverge from the reference policy

#### 3. Observed Symptoms

When `use_kl_loss=True`, I observed the following:

1. **`actor/pg_clipfrac` stays at 0**
   - Suggests that `ratio` always stays within the clipping range

2. **`actor/kl_loss` shows large spikes early, then quickly drops to 0**
   - Indicates rapid policy divergence followed by collapse

3. **Training becomes unstable around step 55-60**
   - Validation metrics (score, format, correctness) drop sharply
   - `response_length` increases dramatically (jumps to ~1000)
   - `critic/rewards` and `critic/advantages` max values drop to 0

4. **Policy generation fails**
   - Unable to generate responses in the correct format
   - Generates extremely long invalid sequences
   - All responses have advantages close to 0

### Suggested Fix

Would it be possible to change line 267 to:

```python
# Suggested fix
policy_loss = policy_loss + kl_loss * self.config.kl_loss_coef
```

### Impact

- All training configurations using `use_kl_loss=True`
- Particularly affects GRPO training (which typically uses KL loss constraint)

### Reproduction Steps

1. Set `actor_rollout_ref.actor.use_kl_loss=True`
2. Run GRPO training
3. Observe training curves:
   - `actor/pg_clipfrac` stays at 0
   - `actor/kl_loss` shows large spikes early
   - Training becomes unstable around steps 50-60
   - Validation metrics drop sharply

### Expected Behavior After Fix

After the fix, you should observe:
- `actor/pg_clipfrac` shows normal fluctuations (not always 0)
- `actor/kl_loss` stays within reasonable range
- Training remains stable without collapse
- Validation metrics improve normally

### Related Code

- `verl/workers/actor/dp_actor.py:267`
- `verl/trainer/ppo/core_algos.py:242-268` (KL penalty calculation)

### Additional Observation

After the policy becomes unstable (around step 55-60), the model generates responses consisting entirely of exclamation marks (`!`). This appears to be a symptom of policy failure:

- The model loses the ability to generate meaningful tokens
- It falls into a degenerate generation pattern
- All generated responses are invalid and consist only of punctuation marks

This symptom, combined with the other observed failures (format score = 0, response length explosion, etc.), suggests that the policy may have broken down, possibly due to the KL loss sign issue.

### References

- PPO Paper: [Proximal Policy Optimization Algorithms](https://arxiv.org/abs/1707.06347)

---

**Note**: Thank you for maintaining this excellent project! I hope this report is helpful. Please let me know if you need any additional information or clarification.
